README
LICENSE

  Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted, provided
that the above copyright notice appears in all copies and that both the
copyright notice and this permission notice appear in supporting
documentation, and that the same name not be used in advertising or
publicity pertaining to distribution of the software without specific,
written prior permission. We make no representations about the
suitability this software for any purpose. It is provided "as is"
without express or implied warranty.

CREDITS

  busybox: BusyBox, licensed under the GNU General Public License version 2.
(http://www.busybox.net/license.html)

  dropbearmulti: Dropbear, distributed under a MIT-style license.
(http://matt.ucc.asn.au/dropbear/dropbear.html)

  rsync: Rsync, licensed under the GNU General Public License (http://www.gnu.org/licenses/gpl.html).
(http://rsync.samba.org)

  sftp-server: OpenSSH, licensed under a BSD licence.
(http://www.openssh.com)

  htop: htop, licensed under the GNU General Public License version 2 (http://www.gnu.org/licenses/old-licenses/gpl-2.0.html).
(http://htop.sourceforge.net/index.php?page=author)

  lsof: lsof, Copyright 2002 Purdue Research Foundation, West Lafayette, Indiana 47907. All rights reserved. Written by Victor A. Abell
(ftp://lsof.itap.purdue.edu/pub/tools/unix/lsof)


ChangeLog.txt
2011-04-19 02:48  NiLuJe

	* [r7420] src/build-updates.sh, src/install.sh, src/uninstall.sh,
	  src/usbnet/info.txt:
	  
	  Kindle Hacks:
	  * USBNetwork:
	  * Tag v0.33.N
	  * Fonts:
	  * Tag v4.5.N
	  * ScreenSavers:
	  * Tag v0.21.N

2011-04-16 03:43  NiLuJe

	* [r7412] src/usbnet/bin/busybox, src/usbnet/bin/dropbearmulti,
	  src/usbnet/bin/htop, src/usbnet/bin/lsof, src/usbnet/bin/rsync,
	  src/usbnet/bin/sftp-server:
	  
	  Kindle Hacks:
	  * Hmm, trusty -O2 builds... :}

2011-04-16 03:09  NiLuJe

	* [r7410] src/usbnet/bin/busybox, src/usbnet/bin/dropbearmulti,
	  src/usbnet/bin/htop, src/usbnet/bin/lsof, src/usbnet/bin/rsync,
	  src/usbnet/bin/sftp-server:
	  
	  Kindle Hacks:
	  * New binaries, new gcc version (still -O3)

2011-04-15 20:23  NiLuJe

	* [r7407] src/usbnet/bin/busybox, src/usbnet/bin/dropbearmulti,
	  src/usbnet/bin/htop, src/usbnet/bin/lsof, src/usbnet/bin/rsync,
	  src/usbnet/bin/sftp-server:
	  
	  Kindle Hacks:
	  * -O3 binaries. So far, it seems slower...

2011-03-29 18:17  NiLuJe

	* [r7387] src/usbnet/bin/rsync:
	  
	  Kindle Hacks:
	  * Fonts:
	  * Update FT snapshot
	  * Update FC snapshot
	  * USBNetwork:
	  * Update rsync
	  * X-TC:
	  * Update ft, fc & rsync.
	  * Let's try again to build FT without -fno-strict-aliasing, and see if
	  it breaks :P

2011-03-23 01:19  NiLuJe

	* [r7379] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-03-23 01:08  NiLuJe

	* [r7377] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-03-23 01:08  NiLuJe

	* [r7376] src/usbnet/bin/usbnet-disable, src/usbnet/bin/usbnet-enable,
	  src/usbnet/bin/usbnetwork:
	  
	  Kindle Hacks:
	  * JailBreak:
	  * Fix the dos2unix sed script to work with busybox's crappy sed
	  implementation
	  * Fonts & ScreenSavers:
	  * Update whitelist
	  * USBNetwork:
	  * Do a dos2unix run on the config file before source'ing it

2011-03-07 21:15  NiLuJe

	* [r7362] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL
	  * Update MR Thread

2011-03-03 17:34  NiLuJe

	* [r7343] src/build-updates.sh, src/install.sh, src/uninstall.sh,
	  src/usbnet/info.txt:
	  
	  Kindle Hacks:
	  * USBNetwork:
	  * Tag v0.32.N
	  * Update MR Thread

2011-03-03 17:33  NiLuJe

	* [r7342] src/usbnet/bin/dropbearmulti:
	  
	  Kindle Hacks:
	  * USBNetwork:
	  * Update dropbear
	  * X-TC:
	  * Bump dropbear to 0.53.1

2011-03-01 15:12  NiLuJe

	* [r7328] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-03-01 15:09  NiLuJe

	* [r7327] src/install.sh:
	  
	  Kindle Hacks:
	  * Fonts & USBNetwork:
	  * Use a more robust method when verifying checksums to detect custom
	  content during install

2011-03-01 14:02  NiLuJe

	* [r7325] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-02-28 19:37  NiLuJe

	* [r7321] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-02-28 16:11  NiLuJe

	* [r7319] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-02-28 15:41  NiLuJe

	* [r7317] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-02-28 14:48  NiLuJe

	* [r7315] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-02-27 20:09  NiLuJe

	* [r7313] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CL

2011-02-27 19:31  NiLuJe

	* [r7312] src/build-updates.sh, src/install.sh, src/uninstall.sh,
	  src/usbnet/info.txt:
	  
	  Kindle Hacks:
	  * Fonts:
	  * Tag v4.3.N
	  * ScreenSavers:
	  * Tag v0.19.N
	  * USBNetwork:
	  * Tag v0.31.N
	  * Update MR Thread

2011-02-27 19:18  NiLuJe

	* [r7311] src/usbnet/bin/busybox, src/usbnet/bin/dropbearmulti,
	  src/usbnet/bin/htop, src/usbnet/bin/rsync, src/usbnet/bin/sftp-server:
	  
	  Kindle Hacks:
	  * Fonts:
	  * Update FreeType libs
	  * Update fc-scan
	  * ScreenSavers:
	  * Update sort
	  * USBNetwork:
	  * Update sftp-server
	  * Update dropbear
	  * Update htop
	  * Update busybox
	  * Update rsync
	  * X-TC:
	  * Update to use the latest versions of a whole bunch of stuff ;)

2010-11-20 00:29  NiLuJe

	* [r7240] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CLs

2010-11-20 00:28  NiLuJe

	* [r7238] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CLs

2010-11-19 22:46  NiLuJe

	* [r7236] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CLs

2010-11-19 02:03  NiLuJe

	* [r7216] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CLs

2010-11-19 02:01  NiLuJe

	* [r7214] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CLs

2010-11-16 04:46  NiLuJe

	* [r7196] ChangeLog:
	  
	  Kindle Hacks:
	  * Udate MR Thread
	  * Update ChangeLogs

2010-11-11 07:33  NiLuJe

	* [r7158] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CLs

2010-11-11 07:31  NiLuJe

	* [r7157] README_FIRST.txt, src/build-updates.sh, src/install.sh,
	  src/uninstall.sh, src/usbnet/bin/usbnetwork, src/usbnet/info.txt:
	  
	  Kindle Hacks:
	  * USBNetwork:
	  * Oops. Drop forgotten unreachable dead code.
	  * Tag v0.30.N

2010-11-11 04:54  NiLuJe

	* [r7156] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CLs

2010-11-11 04:53  NiLuJe

	* [r7155] src/usbnet/bin/usbnetwork:
	  
	  Kindle Hacks:
	  * USBNetwork:
	  * Sigh. Fix the usbnetd call again. I'm stupid.

2010-11-11 04:25  NiLuJe

	* [r7154] ChangeLog:
	  
	  Kindle Hacks:
	  * Update CLs

2010-11-11 04:23  NiLuJe

	* [r7153] src/usbnet/bin/usbnetwork:
	  
	  Kindle Hacks:
	  * USBNet:
	  * Yep. Fix comment.

2010-11-11 04:22  NiLuJe

	* [r7152] src/usbnet/bin/usbnetwork:
	  
	  Kindle Hacks:
	  * USBNet:
	  * Yeah, no, that probablt wouldn't help. Just don't
	  go through rc5 while plugged in...

2010-11-11 03:37  NiLuJe

	* [r7151] README_FIRST.txt, src/install.sh, src/uninstall.sh,
	  src/usbnet-init, src/usbnet/bin/usbnetwork:
	  
	  Kindle Hacks:
	  * USBNetwork:
	  * Fix the "nothing exporte over usbms" issue after an update.

2010-11-11 03:03  NiLuJe


src/uninstall.sh
#!/bin/sh
#
# $Id: uninstall.sh 7343 2011-03-03 17:34:57Z NiLuJe $
#
# diff OTA patch script

_FUNCTIONS=/etc/rc.d/functions
[ -f ${_FUNCTIONS} ] && . ${_FUNCTIONS}


MSG_SLLVL_D="debug"
MSG_SLLVL_I="info"
MSG_SLLVL_W="warn"
MSG_SLLVL_E="err"
MSG_SLLVL_C="crit"
MSG_SLNUM_D=0
MSG_SLNUM_I=1
MSG_SLNUM_W=2
MSG_SLNUM_E=3
MSG_SLNUM_C=4
MSG_CUR_LVL=/var/local/system/syslog_level

logmsg()
{
    local _NVPAIRS
    local _FREETEXT
    local _MSG_SLLVL
    local _MSG_SLNUM

    _MSG_LEVEL=$1
    _MSG_COMP=$2

    { [ $# -ge 4 ] && _NVPAIRS=$3 && shift ; }

    _FREETEXT=$3

    eval _MSG_SLLVL=\${MSG_SLLVL_$_MSG_LEVEL}
    eval _MSG_SLNUM=\${MSG_SLNUM_$_MSG_LEVEL}

    local _CURLVL

    { [ -f $MSG_CUR_LVL ] && _CURLVL=`cat $MSG_CUR_LVL` ; } || _CURLVL=1

    if [ $_MSG_SLNUM -ge $_CURLVL ]; then
        /usr/bin/logger -p local4.$_MSG_SLLVL -t "ota_install" "$_MSG_LEVEL def:$_MSG_COMP:$_NVPAIRS:$_FREETEXT"
    fi

    [ "$_MSG_LEVEL" != "D" ] && echo "ota_install: $_MSG_LEVEL def:$_MSG_COMP:$_NVPAIRS:$_FREETEXT"
}

if [ -z "${_PERCENT_COMPLETE}" ]; then
    export _PERCENT_COMPLETE=0
fi

update_percent_complete()
{
    _PERCENT_COMPLETE=$((${_PERCENT_COMPLETE} + $1))
    update_progressbar ${_PERCENT_COMPLETE}
}

# Hack specific config (name and when to start/stop)
HACKNAME="usbnet"
SLEVEL="82"
KLEVEL="09"

# Based on version 0.11 (20100124), ebs, jya
HACKVER="0.32.N"

# Directories
USBNET_BASEDIR="/mnt/us/usbnet"
USBNET_BINDIR="${USBNET_BASEDIR}/bin"

USBNET_LOG="${USBNET_BASEDIR}/usbnetwork_uninstall.log"

KINDLE_TESTDIR="/test/bin"
KINDLE_USBNETBIN="${KINDLE_TESTDIR}/usbnetwork"

USBNET_USBNETBIN="${USBNET_BINDIR}/usbnetwork"

# Result codes
OK=0
ERR=${OK}

# Here we go
echo >> ${USBNET_LOG}
echo "usbnetwork v${HACKVER}, $( date )" >> ${USBNET_LOG}

update_percent_complete 2

# Remove our deprecated content
# From v0.21.N
logmsg "I" "update" "removing deprecated init scripts & symlinks (v0.21.N)"
if [ -f /etc/init.d/${HACKNAME}-wifi ] ; then
    echo "/etc/init.d/${HACKNAME}-wifi exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/init.d/${HACKNAME}-wifi >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

if [ -L /etc/rc5.d/S72${HACKNAME}-wifi ] ; then
    echo "symbolic link /etc/rc5.d/S72${HACKNAME}-wifi exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rc5.d/S72${HACKNAME}-wifi >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

update_percent_complete 10

# From v0.22.N
logmsg "I" "update" "removing deprecated symlinks (v0.22.N)"
if [ -L /etc/rcS.d/S72${HACKNAME} ] ; then
    echo "symbolic link /etc/rcS.d/S72${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rcS.d/S72${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

# From v0.23.N
logmsg "I" "update" "removing deprecated symlinks (v0.23.N)"
if [ -L /etc/rc5.d/S99${HACKNAME} ] ; then
    echo "symbolic link /etc/rc5.d/S99${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rc5.d/S99${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

update_percent_complete 20

# Remove `usbNetwork command symlink
logmsg "I" "update" "removing usbNetwork private command symlink"
if [ -L ${KINDLE_USBNETBIN} ] ; then
    echo "symbolic link ${KINDLE_USBNETBIN} -> $( readlink ${KINDLE_USBNETBIN} ) exists, deleting..." >> ${USBNET_LOG}
    rm -f ${KINDLE_USBNETBIN} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

update_progressbar 30

# Remove SSH server symlinks
logmsg "I" "update" "removing SSH server symlinks"
LIST="/usr/sbin/dropbearmulti /usr/bin/dropbear /usr/bin/dbclient /usr/bin/dbscp"
for var in ${LIST} ; do
    if [ -L ${var} ] ; then
        echo "symbolic link ${var} -> $( readlink ${var} ) exists, deleting..." >> ${USBNET_LOG}
        DBM=$( readlink ${var} )
        if [ "${DBM}" = "${USBNET_BINDIR}/dropbearmulti" ] ; then
            rm -f ${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
        else
            echo "symbolic link is not ours, skipping..." >> ${USBNET_LOG}
        fi
    fi
done

update_progressbar 40

# Remove lsof symlink
logmsg "I" "update" "removing lsof symlink"
var="/usr/sbin/lsof"
if [ -L ${var} ] ; then
    echo "symbolic link ${var} -> $( readlink ${var} ) exists, deleting..." >> ${USBNET_LOG}
    SYMBIN=$( readlink ${var} )
    if [ "${SYMBIN}" = "${USBNET_BINDIR}/lsof" ] ; then
        rm -f ${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
    else
        echo "symbolic link is not ours, skipping..." >> ${USBNET_LOG}
    fi
fi

update_progressbar 50

# Remove htop symlink
logmsg "I" "update" "removing htop symlink"
var="/usr/bin/htop"
if [ -L ${var} ] ; then
    echo "symbolic link ${var} -> $( readlink ${var} ) exists, deleting..." >> ${USBNET_LOG}
    SYMBIN=$( readlink ${var} )
    if [ "${SYMBIN}" = "${USBNET_BINDIR}/htop" ] ; then
        rm -f ${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
    else
        echo "symbolic link is not ours, skipping..." >> ${USBNET_LOG}
    fi
fi

update_progressbar 60

# Remove rsync symlink
logmsg "I" "update" "removing lsof symlink"
var="/usr/bin/rsync"
if [ -L ${var} ] ; then
    echo "symbolic link ${var} -> $( readlink ${var} ) exists, deleting..." >> ${USBNET_LOG}
    SYMBIN=$( readlink ${var} )
    if [ "${SYMBIN}" = "${USBNET_BINDIR}/rsync" ] ; then
        rm -f ${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
    else
        echo "symbolic link is not ours, skipping..." >> ${USBNET_LOG}
    fi
fi

update_progressbar 70

# Delete init script
logmsg "I" "update" "removing init scripts"
if [ -f /etc/init.d/${HACKNAME} ] ; then
    echo "/etc/init.d/${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/init.d/${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

logmsg "I" "update" "removing boot runlevel symlink"
if [ -L /etc/rc5.d/S${SLEVEL}${HACKNAME} ] ; then
    echo "symbolic link /etc/rc5.d/S${SLEVEL}${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rc5.d/S${SLEVEL}${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

logmsg "I" "update" "removing reboot runlevel symlink"
if [ -L /etc/rc6.d/K${KLEVEL}${HACKNAME} ] ; then
    echo "symbolic link /etc/rc6.d/K${KLEVEL}${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rc6.d/K${KLEVEL}${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

logmsg "I" "update" "removing shutdown runlevel symlink"
if [ -L /etc/rc0.d/K${KLEVEL}${HACKNAME} ] ; then
    echo "symbolic link /etc/rc0.d/K${KLEVEL}${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rc0.d/K${KLEVEL}${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

logmsg "I" "update" "removing update runlevel symlink"
if [ -L /etc/rc3.d/K${KLEVEL}${HACKNAME} ] ; then
    echo "symbolic link /etc/rc3.d/K${KLEVEL}${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rc3.d/K${KLEVEL}${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

update_progressbar 80

# Disable USB network tethering
logmsg "I" "update" "removing traces of USB network tethering"
echo "Part I done! Now to remove traces of USB network tethering..." >> ${USBNET_LOG}

# Handle K3?
for platform in mario luigi ; do
    fwk_cfg="/opt/amazon/ebook/config/framework.${platform}.conf"
    if [[ -f "${fwk_cfg}" ]] ; then
        logmsg I "update" "restoring ${platform} config"
        sed -e "s/USE_WAN : false/USE_WAN : true/g" -i "${fwk_cfg}" >> ${USBNET_LOG} 2>&1 || exit ${ERR}
    fi
done

rm -f /etc/resolv.d/resolv.conf.default >> ${USBNET_LOG} 2>&1 || exit ${ERR}
touch /etc/resolv.d/resolv.conf.default >> ${USBNET_LOG} 2>&1 || exit ${ERR}
cp -f /etc/resolv.d/resolv.conf.default /var/run/resolv.conf >> ${USBNET_LOG} 2>&1 || exit ${ERR}

echo "All done!" >> ${USBNET_LOG}
update_progressbar 90

# From v0.11.N
# Remove custom directory in userstore? [That'll remove the detailed uninstall log, of course]
logmsg "I" "update" "removing custom directory (only if /mnt/us/${HACKNAME}/uninstall exists)"
[ -d /mnt/us/${HACKNAME} -a -f /mnt/us/${HACKNAME}/uninstall ] && rm -rf /mnt/us/${HACKNAME}

logmsg "I" "update" "done"
update_progressbar 100

return ${OK}


src/install.sh
#!/bin/sh
#
# $Id: install.sh 7420 2011-04-19 02:48:56Z NiLuJe $
#
# diff OTA patch script

_FUNCTIONS=/etc/rc.d/functions
[ -f ${_FUNCTIONS} ] && . ${_FUNCTIONS}


MSG_SLLVL_D="debug"
MSG_SLLVL_I="info"
MSG_SLLVL_W="warn"
MSG_SLLVL_E="err"
MSG_SLLVL_C="crit"
MSG_SLNUM_D=0
MSG_SLNUM_I=1
MSG_SLNUM_W=2
MSG_SLNUM_E=3
MSG_SLNUM_C=4
MSG_CUR_LVL=/var/local/system/syslog_level

logmsg()
{
    local _NVPAIRS
    local _FREETEXT
    local _MSG_SLLVL
    local _MSG_SLNUM

    _MSG_LEVEL=$1
    _MSG_COMP=$2

    { [ $# -ge 4 ] && _NVPAIRS=$3 && shift ; }

    _FREETEXT=$3

    eval _MSG_SLLVL=\${MSG_SLLVL_$_MSG_LEVEL}
    eval _MSG_SLNUM=\${MSG_SLNUM_$_MSG_LEVEL}

    local _CURLVL

    { [ -f $MSG_CUR_LVL ] && _CURLVL=`cat $MSG_CUR_LVL` ; } || _CURLVL=1

    if [ $_MSG_SLNUM -ge $_CURLVL ]; then
        /usr/bin/logger -p local4.$_MSG_SLLVL -t "ota_install" "$_MSG_LEVEL def:$_MSG_COMP:$_NVPAIRS:$_FREETEXT"
    fi

    [ "$_MSG_LEVEL" != "D" ] && echo "ota_install: $_MSG_LEVEL def:$_MSG_COMP:$_NVPAIRS:$_FREETEXT"
}

if [ -z "${_PERCENT_COMPLETE}" ]; then
    export _PERCENT_COMPLETE=0
fi

update_percent_complete()
{
    _PERCENT_COMPLETE=$((${_PERCENT_COMPLETE} + $1))
    update_progressbar ${_PERCENT_COMPLETE}
}

# Hack specific config (name and when to start/stop)
HACKNAME="usbnet"
SLEVEL="82"
KLEVEL="09"

# Based on version 0.11 (20100124), ebs, jya
HACKVER="0.33.N"

# Directories
USBNET_BASEDIR="/mnt/us/usbnet"
USBNET_BINDIR="${USBNET_BASEDIR}/bin"

USBNET_LOG="${USBNET_BASEDIR}/usbnetwork_install.log"

KINDLE_TESTDIR="/test/bin"
KINDLE_USBNETBIN="${KINDLE_TESTDIR}/usbnetwork"

USBNET_USBNETBIN="${USBNET_BINDIR}/usbnetwork"

# Result codes
OK=0
ERR=${OK}

update_percent_complete 2

# Install our hack's custom content
# But keep the user's custom content...
if [ -d /mnt/us/${HACKNAME} ] ; then
    logmsg "I" "update" "our custom directory already exists, checking if we have custom content to preserve"
    # Custom IP config
    if [ -f /mnt/us/${HACKNAME}/etc/config ] ; then
        cfg_expected_md5="3d630a6bdbeae41be8c2a2436159eeee 4836fc4a5383b38ff062afebc1da4972 10d30f97f0adbdb86869e48c01095c82"
        cfg_current_md5=$( md5sum /mnt/us/${HACKNAME}/etc/config | awk '{ print $1; }' )
        cfg_md5_match="false"
        for cur_exp_md5 in ${cfg_expected_md5} ; do
            if [ "${cfg_current_md5}" == "${cur_exp_md5}" ] ; then
                cfg_md5_match="true"
            fi
        done
        if [ "${cfg_md5_match}" != "true" ] ; then
            HACK_EXCLUDE="${HACKNAME}/etc/config"
            logmsg "I" "update" "found custom ip config, excluding from archive"
        fi
   fi
fi

update_progressbar 7

# Okay, now we can extract it. Since busybox's tar is very limited, we have to use a tmp directory to perform our filtering
logmsg "I" "update" "installing custom directory"
tar -xvzf ${HACKNAME}.tar.gz

# That's very much inspired from official update scripts ;)
cd src
# And now we filter the content to preserve user's custom content
for custom_file in ${HACK_EXCLUDE} ; do
    if [ -f "./${custom_file}" ] ; then
        logmsg "I" "update" "preserving custom content (${custom_file})"
        rm -f "./${custom_file}"
    fi
done
# Finally, re-tape our filtered dir and unleash it on the live userstore
tar cf - . | (cd /mnt/us ; tar xvf -)
_RET=$?
if [ ${_RET} -ne 0 ] ; then
    logmsg "C" "update" "code=${_RET}" "failure to update userstore with custom directory"
    return 1
fi
cd - >/dev/null
rm -rf src

update_progressbar 14

# Here we go
echo >> ${USBNET_LOG}
echo "usbnetwork v${HACKVER}, $( date )" >> ${USBNET_LOG}

update_progressbar 21

# Remove our deprecated content
# From v0.21.N
logmsg "I" "update" "removing deprecated init scripts & symlinks (v0.21.N)"
if [ -f /etc/init.d/${HACKNAME}-wifi ] ; then
    echo "/etc/init.d/${HACKNAME}-wifi exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/init.d/${HACKNAME}-wifi >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

if [ -L /etc/rc5.d/S72${HACKNAME}-wifi ] ; then
    echo "symbolic link /etc/rc5.d/S72${HACKNAME}-wifi exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rc5.d/S72${HACKNAME}-wifi >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

update_percent_complete 24

# From v0.22.N
logmsg "I" "update" "removing deprecated symlinks (v0.22.N)"
if [ -L /etc/rcS.d/S72${HACKNAME} ] ; then
    echo "symbolic link /etc/rcS.d/S72${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rcS.d/S72${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

# From v0.23.N
logmsg "I" "update" "removing deprecated symlinks (v0.23.N)"
if [ -L /etc/rc5.d/S99${HACKNAME} ] ; then
    echo "symbolic link /etc/rc5.d/S99${HACKNAME} exists, deleting..." >> ${USBNET_LOG}
    rm -f /etc/rc5.d/S99${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

update_progressbar 28

# Make sure our custom binaries are executable
LIST="usbnetwork busybox dropbearmulti rsync usbnet-enable usbnet-disable sftp-server htop lsof"
for var in ${LIST} ; do
    [ -x ${USBNET_BINDIR}/${var} ] || chmod +x ${USBNET_BINDIR}/${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
done

update_progressbar 35

# Make sure the /test/bin directory exists
logmsg "I" "update" "creating the /test/bin directory if need be"
[ -d ${KINDLE_TESTDIR} ] || mkdir -p ${KINDLE_TESTDIR} >> ${USBNET_LOG} 2>&1 || exit ${ERR}

update_progressbar 42

# Setup SSH server
logmsg "I" "update" "installing SSH server"
LIST="/usr/sbin/dropbearmulti /usr/bin/dropbear /usr/bin/dbclient /usr/bin/dbscp"
for var in ${LIST} ; do
    if [ -L ${var} ] ; then
        echo "symbolic link ${var} -> $( readlink ${var} ) already exists, skipping..." >> ${USBNET_LOG}
    else
        if [ -x ${var} ] ; then
            echo "Binary ${var} already exists, skipping..." >> ${USBNET_LOG}
        else
            ln -fs ${USBNET_BINDIR}/dropbearmulti ${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
        fi
    fi
done

update_progressbar 49

# Setup lsof
logmsg "I" "update" "installing lsof"
var="/usr/sbin/lsof"
if [ -L ${var} ] ; then
    echo "symbolic link ${var} -> $( readlink ${var} ) already exists, skipping..." >> ${USBNET_LOG}
else
    if [ -x ${var} ] ; then
        echo "Binary ${var} already exists, skipping..." >> ${USBNET_LOG}
    else
        ln -fs ${USBNET_BINDIR}/lsof ${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
    fi
fi

update_progressbar 56

# Setup htop
logmsg "I" "update" "installing htop"
var="/usr/bin/htop"
if [ -L ${var} ] ; then
    echo "symbolic link ${var} -> $( readlink ${var} ) already exists, skipping..." >> ${USBNET_LOG}
else
    if [ -x ${var} ] ; then
        echo "Binary ${var} already exists, skipping..." >> ${USBNET_LOG}
    else
        ln -fs ${USBNET_BINDIR}/htop ${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
    fi
fi

update_progressbar 63

# Setup rsync
logmsg "I" "update" "installing rsync"
var="/usr/bin/rsync"
if [ -L ${var} ] ; then
    echo "symbolic link ${var} -> $( readlink ${var} ) already exists, skipping..." >> ${USBNET_LOG}
else
    if [ -x ${var} ] ; then
        echo "Binary ${var} already exists, skipping..." >> ${USBNET_LOG}
    else
        ln -fs ${USBNET_BINDIR}/rsync ${var} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
    fi
fi

update_progressbar 70

# Setup `usbNetwork command symlink
logmsg "I" "update" "setting up usbNetwork private command symlink"
if [ -L ${KINDLE_USBNETBIN} ] ; then
    echo "symbolic link ${KINDLE_USBNETBIN} -> $( readlink ${KINDLE_USBNETBIN} ) already exists, skipping..." >> ${USBNET_LOG}
else
    # Save normal file in case it already exists
    if [ -f ${KINDLE_USBNETBIN} ] ; then
        echo "${KINDLE_USBNETBIN} exists, saving..." >> ${USBNET_LOG}
        cp ${KINDLE_USBNETBIN} ${USBNET_USBNETBIN}-save.${HACKVER} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
        rm -f ${KINDLE_USBNETBIN} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
    fi

    # Create a link
    ln -fs ${USBNET_USBNETBIN} ${KINDLE_USBNETBIN} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

update_progressbar 77

# Setup auto USB network startup script
logmsg "I" "update" "installing init script"
cp -f ${HACKNAME}-init /etc/init.d/${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
[ -x /etc/init.d/${HACKNAME} ] || chmod +x /etc/init.d/${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}

update_progressbar 84

# And make it start at boot, after userstore, after network, after volumd, and before framework (rc5)
logmsg "I" "update" "creating boot symlink"
if [ -L /etc/rc5.d/S${SLEVEL}${HACKNAME} ] ; then
    echo "symbolic link /etc/rc5.d/S${SLEVEL}${HACKNAME} already exists, skipping..." >> ${USBNET_LOG}
else
    ln -fs /etc/init.d/${HACKNAME} /etc/rc5.d/S${SLEVEL}${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

# Make it stop at reboot (rc6), after the framework and before userstore
logmsg "I" "update" "creating reboot runlevel symlink"
if [ -L /etc/rc6.d/K${KLEVEL}${HACKNAME} ] ; then
    echo "symbolic link /etc/rc6.d/K${KLEVEL}${HACKNAME} already exists, skipping..." >> ${USBNET_LOG}
else
    ln -fs /etc/init.d/${HACKNAME} /etc/rc6.d/K${KLEVEL}${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

# Make it stop at shutdown (rc0), after the framework and before userstore
logmsg "I" "update" "creating shutdown runlevel symlink"
if [ -L /etc/rc0.d/K${KLEVEL}${HACKNAME} ] ; then
    echo "symbolic link /etc/rc0.d/K${KLEVEL}${HACKNAME} already exists, skipping..." >> ${USBNET_LOG}
else
    ln -fs /etc/init.d/${HACKNAME} /etc/rc0.d/K${KLEVEL}${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

# Make it stop when updating (rc3), after the framework and before the updater
logmsg "I" "update" "creating update runlevel symlink"
if [ -L /etc/rc3.d/K${KLEVEL}${HACKNAME} ] ; then
    echo "symbolic link /etc/rc3.d/K${KLEVEL}${HACKNAME} already exists, skipping..." >> ${USBNET_LOG}
else
    ln -fs /etc/init.d/${HACKNAME} /etc/rc3.d/K${KLEVEL}${HACKNAME} >> ${USBNET_LOG} 2>&1 || exit ${ERR}
fi

update_progressbar 91

# Cleanup
logmsg "I" "update" "cleaning up"
rm -f ${HACKNAME}-init ${HACKNAME}.tar.gz

update_progressbar 98

echo "Done!" >> ${USBNET_LOG}
logmsg "I" "update" "done"
update_progressbar 100

return ${OK}
